Source/SuperLinq/IndexOf.cs

```diff
@@ -0,0 +1,123 @@
+	/// cref="IEnumerable{T}"/>, if found; otherwise, <c>-1</c>.
+	/// </returns>
+	/// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+	/// <remarks>
+	/// <para>
+	/// The <see cref="IEnumerable{T}"/> is searched forward starting at the first element and ending at the last
+	/// element.
+	/// </para>
+	/// <para>
+	/// This method determines equality using the default equality comparer <see cref="EqualityComparer{T}.Default"/>
+	/// for <typeparamref name="TSource"/>, the type of values in the list.
+	/// </para>
+	/// <para>
+	/// This operator executes immediately.
+	/// </para>
+	/// </remarks>
+	public static int IndexOf<TSource>(this IEnumerable<TSource> source, TSource item)
+	{
+		return source.IndexOf(item, 0);
+	}
+
+	/// <summary>
+	/// Searches for the specified object and returns the zero-based index of the first occurrence within the range of
+	/// elements in the <see cref="IEnumerable{T}"/> that extends from the specified index to the last element.
+	/// </summary>
+	/// <typeparam name="TSource">
+	/// The type of elements of <paramref name="source"/></typeparam>
+	/// <param name="source">The source sequence.</param>
+	/// <param name="item">The object to locate in the <see cref="IEnumerable{T}"/>. The value can be <see
+	/// langword="null"/> for reference types.</param>
+	/// <param name="index">The zero-based starting index of the search. <c>0</c> (zero) is valid in an empty
+	/// list.</param>
+	/// <returns>
+	/// The zero-based index of the first occurrence of <paramref name="item"/> within the the range of elements in the
+	/// <see cref="IEnumerable{T}"/> that extends from <paramref name="index"/> to the last element, if found;
+	/// otherwise, <c>-1</c>.
+	/// </returns>
+	/// <exception cref="ArgumentNullException"><paramref name="source"/> is null.</exception>
+	/// <exception cref="ArgumentOutOfRangeException"><paramref name="index"/> is less than <c>0</c>.</exception>
+	/// <remarks>
+	/// <para>
+	/// The <see cref="IEnumerable{T}"/> is searched forward starting at <paramref name="index"/> and ending at the last
+	/// element.
+	/// </para>
+	/// <para>
+	/// This method determines equality using the default equality comparer <see cref="EqualityComparer{T}.Default"/>
+	/// for <typeparamref name="TSource"/>, the type of values in the list.
+	/// </para>
+	/// <para>
+	/// This operator executes immediately.
+	/// </para>
+	/// </remarks>
+	public static int IndexOf<TSource>(this IEnumerable<TSource> source, TSource item, Index index)
+	{
+		Guard.IsNotNull(source);
+
+		if (TryGetCollectionCount(source, out var length))
+		{
+			index = index.GetOffset(length);
+		}
+
+		if (!index.IsFromEnd)
+		{
+			var i = 0;
+			foreach (var element in source)
+			{
+				if (i >= index.Value && EqualityComparer<TSource>.Default.Equals(element, item))
+					return i;
+
+				i++;
+			}
+
+			return -1;
+		}
+		else
+		{
```

Comment:
You can add an optimization for `I(ReadOnly)List` by enumerating backwards allocation-free

Reply:
(This holds for all of #109)